### PR TITLE
Add threat model for the CLI

### PR DESCRIPTION
**Pull request by st3penta**
Add `THREAT_MODEL.md` at the repo root following the wg-agentic-sdlc 8-section schema. Based on the AI-generated draft attached to EC-2001, validated and corrected against the codebase (entry points, flag semantics, merge behavior, dependency references).

Covers all three operating modes (CLI, server, acceptance test) with entry points, trust boundaries, and threat categories.

Ref: https://issues.redhat.com/browse/EC-2001

Co-Authored-By: Claude Opus 4.6 <noreply@anthropic.com>

### CODE REVIEW

File: THREAT_MODEL.md [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+| `--vsa-public-key` | Operator-controlled | Public key used to verify VSA signatures. File path or K8s Secret reference. |
+| `--fallback-public-key` | Operator-controlled | Public key used for fallback image validation when VSA validation fails. This is a different key from the VSA verification key, used for a fundamentally different validation mode (full image signature verification). |
+| `--no-fallback` | Operator-controlled | Disables fallback to full image validation when VSA validation fails. When set, a VSA validation failure is final with no image validation retry. Fallback is enabled by default. |
+| `--ignore-signature-verification` | Operator-controlled | Skips VSA signature verification entirely. Parallels `--skip-att-sig-check` for attestations. |
+| `--vsa-retrieval` | Operator-controlled | Controls how VSAs are retrieved (e.g., from Rekor or local storage). |
+
+The `--ignore-signature-verification` flag is a direct bypass mechanism for VSA
+signature checking. When enabled, unverified VSAs are accepted, making threats
+CA-6 (VSA-based validation skip) and CR-2 (VSA signing key compromise) moot
+since signature verification is not performed at all.
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md]: Good catch. The flag removes the verification control rather than eliminating the threat. Reworded to describe the actual risk and note that deployments relying on VSA trust should restrict this flag.

File: THREAT_MODEL.md [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+| NW-1 | **MITM on registry traffic**: attacker intercepts OCI registry connections to serve modified policy bundles, data sources, or attestations. | Critical -- attacker controls verification inputs. | Low -- all OCI registry traffic uses TLS. | `isSecure()` rejects plaintext HTTP sources. TLS certificate validation via Go's standard library. |
+| NW-2 | **Registry unavailability**: OCI registry or sigstore infrastructure is unavailable, preventing validation from completing. | Medium -- build/release pipeline stalls. | Medium -- depends on registry/service uptime. | Retry with exponential backoff (via `oras-go` retry transport). Timeout configuration. |
+| NW-3 | **DNS rebinding in server mode**: if the server binds to `0.0.0.0`, it is accessible from any network. An attacker on the same network could submit evaluation requests. | Medium -- unauthorized policy evaluation, potential information disclosure of policy configuration. | Low -- default binding is `127.0.0.1`. Documentation warns about production use without a reverse proxy. | Default localhost binding. Server documentation recommends reverse proxy or network policy for production. |
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md]: Right, that's unauthenticated network exposure, not DNS rebinding. Renamed and adjusted the description.

File: THREAT_MODEL.md [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+| RT-2 | **Stale policies in server mode**: policies are loaded once at startup. Policy changes require a server restart. A long-running server may enforce outdated policies. | Medium -- new policy rules or data are not applied until restart. | Medium -- depends on operational procedures. | Documented behavior: "Restart the server to pick up policy changes." |
+| RT-3 | **Temp file leakage**: server-mode input files and `ec-work-*` directories may contain sensitive data (attestation content, policy configuration). | Low -- temp files are cleaned up after each request/evaluation. | Low -- files are created with default permissions, removed on completion. `--debug` mode preserves work directories. | `defer os.Remove(tmpPath)` in handler. Config files written with 0444 permissions. |
+| RT-4 | **Panic recovery exposes internal state**: the recovery middleware catches panics and returns a generic error, but stack traces are logged. | Low -- stack traces go to server logs, not to the HTTP response. | Low -- panics are exceptional. | Recovery middleware returns generic "internal server error". Stack trace logged at error level (not returned to caller). |
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md]: Fair point. 0444 is world-readable, so listing it as a mitigation is misleading. Updated to describe the actual behavior and reference the mitigation recommendation for 0600.

File: THREAT_MODEL.md [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+
+
+## 8. Recommended Mitigations
+
+### High priority
+
+1. **Report signature skip status**: add a field to the JSON output report
+   indicating whether image signature and/or attestation signature verification
+   was skipped. This makes the `--skip-*-sig-check` flags visible to downstream
+   consumers and prevents silent bypass. (Addresses CA-3, open question 2.)
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md]: Agreed. Visibility enables detection, not prevention. Clarified the wording.

File: THREAT_MODEL.md [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+- **CLI mode** (primary): one-shot evaluation invoked by CI pipelines and
+  developers. The main commands are `ec validate image` (verifies container
+  images against policy), `ec validate input` (verifies arbitrary JSON/YAML
+  against policy), and `ec validate vsa` (validates Verification Summary
+  Attestations). Additional subcommands: `ec validate policy` (validate policy
+  bundles), `ec fetch`, `ec inspect`, `ec opa`, `ec sigstore`, `ec track`,
+  `ec compare`.
+- **Server mode**: a persistent HTTP server (`ec validate input --server`) that
+  exposes a REST API for policy evaluation. Policies are loaded once at startup;
+  the server accepts input documents over HTTP and returns evaluation results.
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md]: Checked the code. NewInput constructs evaluators and initializes work/data directories, but policy bundle downloads happen on first Evaluate call (cached via sync.OnceValues). Fixed the description and /ready entry to reflect when things actually load.

File: THREAT_MODEL.md [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+| RT-1 | **Denial of service via large input**: server mode accepts up to 80MB request bodies. Concurrent large requests could exhaust memory or disk. | Medium -- server instance becomes unresponsive. | Medium -- no rate limiting or concurrent request limits. | 80MB body size limit. 90-second evaluation timeout. Temp files cleaned up after each request. |
+| RT-2 | **Stale policies in server mode**: policies are loaded once at startup. Policy changes require a server restart. A long-running server may enforce outdated policies. | Medium -- new policy rules or data are not applied until restart. | Medium -- depends on operational procedures. | Documented behavior: "Restart the server to pick up policy changes." |
+| RT-3 | **Temp file leakage**: server-mode input files and `ec-work-*` directories may contain sensitive data (attestation content, policy configuration). | Low -- temp files are cleaned up after each request/evaluation. | Low -- files are created with default permissions, removed on completion. `--debug` mode preserves work directories. | `defer os.Remove(tmpPath)` in handler. Config files written with 0444 permissions. |
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md]: Correct. Work directory retention is controlled by the EC_DEBUG environment variable, not the --debug CLI flag. Fixed.

File: THREAT_MODEL.md [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+| SC-2 | **Compromised data source**: attacker modifies external data source content (e.g., trusted task list, allowed registry prefixes) to weaken policy enforcement. | High -- specific rules are weakened. | Medium -- data sources may be in repos with broader write access than the policy repo. | Data sources are fetched over TLS. Content is merged into OPA data tree but existing keys cannot be overridden (only new keys added). |
+| SC-3 | **Dependency compromise**: a transitive dependency (cosign, OPA, go-containerregistry, etc.) is compromised, introducing vulnerabilities in signature verification or policy evaluation. | Critical -- undermines core verification logic. | Low -- major open-source projects with active security review. | Renovate for dependency updates. `go.sum` integrity verification. `CGO_ENABLED=0` reduces native dependency surface. |
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md]: Checked the merge implementation (equivalence.go:278). Non-map keys at the same path are overwritten by the later source (dst[k] = v). Only nested map keys are recursively merged. Updated SC-2 and the data source deep merge entry to reflect the actual behavior.

File: THREAT_MODEL.md [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+| SBOM fetch | Untrusted (attacker-controlled image) | Software Bill of Materials fetched via custom OPA builtins (`ec.oci.*`). Parsed by CycloneDX/SPDX libraries. |
+| VSA fetch | Semi-trusted | VSAs retrieved from Rekor or local storage. Must be signature-verified and within expiration window. |
+
+### 3.3 Sigstore infrastructure
+
+| Entry point | Trust level | Description |
+|-------------|-------------|-------------|
+| Cosign verification | Trusted infrastructure | Signature verification via cosign libraries. Relies on correct key/certificate configuration. |
+| Rekor transparency log | Trusted infrastructure | Transparency log queries/uploads. Subject to network availability and TLS. |
+| Fulcio CA / TUF root | Trusted infrastructure | Certificate authority and trust root for keyless verification. |
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md:119]: Clarified. The pipeline sets it true for traditional signing, but the task parameter defaults to false and keyless verification forces it to false.

File: THREAT_MODEL.md [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+| `POST /v1/validate/input` | Network caller | Accepts JSON/YAML body (up to 80MB). Written to temp file, evaluated against pre-loaded policies. |
+| `GET /live` | Network caller | Liveness probe. Always returns 200. |
+| `GET /ready` | Network caller | Readiness probe. Returns 200 when policies are loaded. |
+
+Server mode has **no built-in authentication, authorization, or rate limiting**.
+The server binds to `127.0.0.1` by default but can be configured to `0.0.0.0`.
+
+### 3.6 Custom OPA builtins
+
+The CLI registers custom OPA builtins that extend the Rego evaluation
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md:143]: Added a parenthetical clarifying these are ec validate image flags.

File: THREAT_MODEL.md [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+| Kubernetes API server | Resolve ECP custom resources, fetch public keys from Secrets |
+| Git hosting (GitHub, GitLab) | Download policy configuration from git repos |
+| OPA/Conftest engine (embedded) | Evaluate Rego policies against attestation data |
+
+Key direct dependencies: `sigstore/cosign`, `sigstore/rekor`,
+`sigstore/sigstore`, `open-policy-agent/opa`, `open-policy-agent/conftest`,
+`google/go-containerregistry` (via `conforma/go-containerregistry` fork),
+`in-toto/in-toto-golang`, `conforma/go-gather`, `conforma/crds`.
+
+
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md]: Fixed, single blank lines throughout.

File: THREAT_MODEL.md [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+| **OPA/Conftest engine vulnerabilities** | OPA is a mature, widely-deployed policy engine with active security review. Bugs are possible but are upstream concerns, not specific to the CLI's threat model. The CLI tracks upstream releases via Renovate. |
+| **Operator misconfiguration** | Operators who misconfigure the ECP (wrong key, permissive exclusions, etc.) introduce risk, but this is an operational concern, not a software vulnerability. The CLI faithfully enforces whatever policy it is given. |
+| **Physical access to build infrastructure** | Out of scope. Assumes standard datacenter/cloud physical security. |
+| **Side-channel attacks on cryptographic operations** | The CLI uses Go's standard crypto libraries and cosign. Side-channel resistance is an upstream concern. The CLI does not implement custom cryptography. |
+| **Concurrent evaluation race conditions** | The worker pool model uses channels for communication. Workers operate on independent components with component-scoped caches. No shared mutable state between workers. |
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md:247]: The deprioritization still holds -- those are mutex-protected caches, not unguarded shared state. But the rationale wording is imprecise. Not worth a code change for a documentation nuance in the deprioritized section.

File: THREAT_MODEL.md [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+
+9. **Server mode policy reload**: add a mechanism to reload policies without
+   restarting the server (e.g., SIGHUP handler, reload endpoint), so that
+   long-running servers can pick up policy changes. (Addresses RT-2.)
+
+10. **Temp file permissions**: ensure server-mode temp files are created with
+    restrictive permissions (0600) rather than relying on umask. (Addresses
+    RT-3.)
+
+11. **Document custom builtin network access**: clearly document that
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md:373]: Good catch on the specificity. os.CreateTemp already uses 0600, so the recommendation should target evaluator config files specifically. Minor precision issue that doesn't change the mitigation direction.

File: THREAT_MODEL.md [hunk 13]
```diff
@@ -0,0 +1,376 @@
+# Threat Model: conforma/cli
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md:1]: Out of scope for this PR. AGENTS.md changes can follow separately if needed.

File: THREAT_MODEL.md [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+   bearer token) or clearly document that a reverse proxy with authentication
+   is required. (Addresses NW-3, RT-1, open question 3.)
+
+6. **Rate limiting in server mode**: add configurable rate limiting and
+   concurrent request limits to prevent resource exhaustion from large or
+   frequent requests. (Addresses RT-1.)
+
+7. **`EXTRA_RULE_DATA` auditing**: log or report what extra rule data was
+   injected, so that release pipeline audits can detect unexpected data
+   manipulation. (Addresses CA-4, open question 6.)
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md:357]: Extra-rule-data contains policy thresholds and allowed lists, not secrets. The recommendation is about audit visibility (detecting unexpected manipulation), not dumping raw values. Logging which keys were injected is enough for the audit use case.

File: THREAT_MODEL.md [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+
+### 3.1 CLI arguments and flags
+
+| Entry point | Trust level | Description |
+|-------------|-------------|-------------|
+| `--policy` / `-p` | Varies | ECP reference: K8s name, file path, git URL, or inline JSON. In integration tests, user-controlled. In release gate, SRE-controlled. |
+| `--image` / `--images` | User-controlled | Image references or Snapshot spec. Determines what is validated. |
+| `--public-key` / `-k` | Varies | Overrides the key from ECP. In Konflux, set by the Tekton task (from K8s Secret). |
+| `--rekor-url` / `-r` | Varies | Rekor URL for transparency log checks. In Konflux, set by `collect-keyless-params`. |
+| `--ignore-rekor` | Varies | Disables Rekor transparency log checks. Set to `true` in Konflux pipelines for traditional signing verification. Forced to `false` for keyless verification. |
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md]: Good catch. The CLI sets IgnoreTlog unconditionally from the flag value (policy.go:535). The keyless enforcement happens at the pipeline level via collect-keyless-params, not in the CLI. Reworded to make that distinction clear.

File: THREAT_MODEL.md [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+| RT-2 | **Stale policies in server mode**: policies are loaded once at startup. Policy changes require a server restart. A long-running server may enforce outdated policies. | Medium -- new policy rules or data are not applied until restart. | Medium -- depends on operational procedures. | Documented behavior: "Restart the server to pick up policy changes." |
+| RT-3 | **Temp file leakage**: server-mode input files and `ec-work-*` directories may contain sensitive data (attestation content, policy configuration). | Low -- temp files are cleaned up after each request/evaluation. | Low -- files are created with default permissions, removed on completion. Setting `EC_DEBUG` env var preserves work directories for inspection. | `defer os.Remove(tmpPath)` in handler. Config files written with 0444 (world-readable) permissions; see mitigation #10 for the recommendation to restrict to 0600. |
```
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md:234]: Already addressed in the previous round. RT-3 now explicitly says 0444 is world-readable and references mitigation #10. The mitigation targets evaluator config files specifically.

File: THREAT_MODEL.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (st3penta) [reply THREAT_MODEL.md:1]: The repo slug is unambiguous for a per-repo threat model. Keeping as-is.

**Review by joejstuart — APPROVED** [review]
